internal/serverinstall/k8s.go

```diff
@@ -371,6 +375,263 @@ func (i *K8sInstaller) Install(
 	}, nil
 }
 
+// Upgrade is a method of K8sInstaller and implements the Installer interface to
+// upgrade a waypoint-server in a Kubernetes cluster
+func (i *K8sInstaller) Upgrade(
+	ctx context.Context, opts *InstallOpts, serverCfg serverconfig.Client) (
+	*InstallResults, error,
+) {
+	ui := opts.UI
+	log := opts.Log
+
+	sg := ui.StepGroup()
+	defer sg.Wait()
+
+	s := sg.Add("Inspecting Kubernetes cluster...")
+	defer s.Abort()
+
+	// Build our K8S client.
+	newCmdConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		clientcmd.NewDefaultClientConfigLoadingRules(),
+		&clientcmd.ConfigOverrides{},
+	)
+
+	// Discover the current target namespace in the user's config so if they
+	// run kubectl commands waypoint will show up. If we use the default namespace
+	// they might not see the objects we've created.
+	if i.config.namespace == "" {
+		namespace, _, err := newCmdConfig.Namespace()
+		if err != nil {
+			ui.Output(
+				"Error getting namespace from client config: %s", clierrors.Humanize(err),
+				terminal.WithErrorStyle(),
+			)
+			return nil, err
+		}
+		i.config.namespace = namespace
+	}
+
+	clientconfig, err := newCmdConfig.ClientConfig()
+	if err != nil {
+		ui.Output(
+			"Error initializing kubernetes client: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(clientconfig)
+	if err != nil {
+		ui.Output(
+			"Error initializing kubernetes client: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return nil, err
+	}
+	// Do some probing to see if this is OpenShift. If so, we'll switch the config for the user.
+	// Setting the OpenShift flag will short circuit this.
+	if !i.config.openshift {
+		s.Update("Gathering information about the Kubernetes cluster...")
+
+		namespaceClient := clientset.CoreV1().Namespaces()
+		_, err := namespaceClient.Get(context.TODO(), "openshift", metav1.GetOptions{})
+		isOpenShift := err == nil
+
+		// Default namespace in OpenShift acts like a regular K8s namespace, so we don't want
+		// to remove fsGroup in this case.
+		if isOpenShift && i.config.namespace != "default" {
+			s.Update("OpenShift detected. Switching configuration...")
+			i.config.openshift = true
+		}
+	}
+
+	s.Done()
+
+	statefulSetClient := clientset.AppsV1().StatefulSets(i.config.namespace)
+	waypointStatefulSet, err := statefulSetClient.Get(ctx, serverName, metav1.GetOptions{})
+	if err != nil {
+		ui.Output(
+			"Error obtaining waypoint statefulset: %s", clierrors.Humanize(err),
+			terminal.WithErrorStyle(),
+		)
+		return nil, err
+	}
+
+	s = sg.Add("Upgrading server to %q", i.config.serverImage)
+
+	// Update pod image to requested serverImage
+	cmd := exec.Command("kubectl", "set", "image", "statefulset", serverName, fmt.Sprintf("server=%s", i.config.serverImage))
```

Comment:
This has to be changed to use the K8S API directly, I don't think we assume `kubectl` is installed OR configured properly anywhere else [anymore].

Reply:
Ah, I was afraid of that! So I noticed on our Install we use `kubectl`: https://github.com/hashicorp/waypoint/blob/6f6b1fb588c5514711bc542447b65a902e85ee60/internal/serverinstall/k8s.go#L156-L161

And was hoping maybe we could keep using it.

Otherwise, my plan was essentially to do exactly what `kubectl set image` does internally but for a single patch updating the pod image: https://github.com/kubernetes/kubectl/blob/master/pkg/cmd/set/set_image.go#L231

Seeing how `kubectl` handles updating k8s got me thinking if we need some kind of similar mechanism for updating k8s in Waypoint? Should we have an internal Patch function like this for future config updates and Upgrade just uses that here?

Reply:
I think I can actually patch the pod itself (like how I'm deleting pods for OnDelete) rather than Statefulset, which I think makes this a bit easier. I'll give that a shot instead.